cedar-policy-validator/src/types.rs

```diff
@@ -627,7 +630,163 @@ impl Type {
         }
     }
 
+    /// Does the given `PartialValue` have this validator type?
+    ///
+    /// If the `PartialValue` is a residual with not enough information to
+    /// determine conclusively that it either does or does not typecheck (i.e.,
+    /// it does typecheck for some permissible substitutions of the unknowns,
+    /// but does not typecheck for other permissible substitutions), this is
+    /// reported as an error.
+    ///
+    /// TODO(#437): Handling of `Unknown`s is not yet complete and doesn't
+    /// properly behave according to the above description, as of this writing.
+    pub(crate) fn typecheck_partial_value(
+        &self,
+        value: &PartialValue,
+        extensions: Extensions<'_>,
+    ) -> Result<bool, GetSchemaTypeError> {
```

Comment:
The conversion to `RestrictedExpr` is fallible, because of the case of nontrivial residuals (residuals which are more than just a single Unknown).  Eventually we will need to handle the nontrivial-residuals case in a better way.

Reply:
Looking at this again, I do think we can reduce the code duplication some though.  I think the appropriate place to delegate is to have `typecheck_value()` convert the `Value` to `RestrictedExpr` and then reuse `typecheck_restricted_expr()`.  That conversion is infallible, and while not free (requires cloning records, among other things), it's probably worth it to avoid the maintenance burden

Reply:
Refactored